### PR TITLE
Stabilize `--env-set` option

### LINKED ISSUE #80792: Add option to pass environment variables to rustc with a flag
See #78913 for more background.

I'm a meson developer, working to make rust work better with meson. I know that cargo is the official build system for rust, but for me and the projects I work on cargo is unacceptable, we have a *massive* C/C++ codebase already using meson, with interest in migrating some (but not all) of our code to rust. Mostly things work fine, but one problem we do have is including generated code.

Cargo handles generated code by setting an `OUT_DIR` environment variable, this works for cargo because it has a number of unique design decisions compared to most build systems. For Cmake and meson and particular this is problematic because we generate a declarative build system which the user invokes, visual-studio, ninja, make, etc; so there are separate "configure" and "build" stages. For ninja in particular environment variables are problematic because it's made an explicit design choice not to support them. As such, we have no way to set environment variables at build time without resorting to something ugly like wrapping rustc in a script to proxy the variables.

What would be nice is the ability to pass environment variables to rustc directly, much like `--cfg`, something like `--env "OUT_DIR=/some/dir"`. This would allow us to make `include()` work like cargo, but without needing to resort to wrapping rustc and making things slower than they need to be.

### LINKED ISSUE #118372: Tracking issue for `--env-set` flag
Adds a new `--env-set` command line option on rustc. It allows to set environment variables value at compile-time to be used by `env!` and `option_env!` macros.

When retrieving and environment variable value, the one specified by `--env-set` will take precedence. For example if you want have `PATH=a` in your environment and pass `--env-set PATH=env`, then you will have:

```rust
assert_eq!(env!("PATH"), "env");
```

### Steps

- [x] Implement the [MCP](https://github.com/rust-lang/compiler-team/issues/653) -> Done in https://github.com/rust-lang/rust/pull/118368, in https://github.com/rust-lang/rust/pull/118830 and in https://github.com/rust-lang/rust/pull/119884
- [ ] Stabilization PR - https://github.com/rust-lang/rust/pull/119926

### Unresolved questions

- [ ] What is the desired behavior if the same `--env` variable is passed twice?

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #118372.
Fixes https://github.com/rust-lang/rust/issues/80792.

As discussed with meson devs, the option by itself is good enough to be used, so there isn't really a need to wait for https://github.com/rust-lang/rfcs/pull/2794 to stabilize `--env-set`.

# Summary

This PR stabilizes the `--env-set` option. This option flag allows specification of environment variables at compile time for the `env!` and `option_env!` macros and `tracked_env::var` function from the `proc_macro` crate.

This information will be stored in the dep-info files. For more information about dep-info files, take a look [here](https://doc.rust-lang.org/cargo/guide/build-cache.html#dep-info-files).

When retrieving an environment variable, the value specified by `--env-set` will take
precedence. For example, if you want have `PATH=env` in your environment and pass:

```bash
rustc --env-set PATH=env
```

Then you will have:

```rust,no_run
assert_eq!(env!("PATH"), "env");
```

Crates will be fully re-compiled if `--env-set` arguments are changed.

Tracking issue is [#118372](https://github.com/rust-lang/rust/issues/118372) (it contains le list of PRs for implementing it).

# Motivation

External tools like `ninja` or `meson` need to be able to specify environment variables through the command line to override current process environment. It also allows to force a recompilation if an environment variable passed with this flag is modified.

# Use of this feature

It is currently used in https://github.com/mesonbuild/meson/pull/10030.

## Tests

List of tests for this option flag:

 * `tests/ui/extenv/extenv-env-overload.rs`
 * `tests/ui/extenv/extenv-env.rs`
 * `tests/ui/extenv/extenv-not-env.rs`
 * `tests/ui/feature-gates/env-flag.rs`
 * `tests/ui/proc-macro/auxiliary/env.rs`
 * `tests/ui/proc-macro/env.rs`

r? @davidtwco

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Can you link to some discussion confirming that ninja has integrated this flag experimentally and that it works properly? We definitely need confirmation for that first before stabilizing anything.

**Comment by GuillaumeGomez** [comment]
ninja? Also we can postpone approving this PR. No hurry here.

**Comment by Nilstrieb** [comment]
I thought this flag was added for ninja

**Comment by GuillaumeGomez** [comment]
Ah I see. No idea. "meson devs" so not sure for what they use it exactly.

**Comment by Nilstrieb** [comment]
Marking it as waiting on author, you should prepare a stabilization report mentioning the use case again and pointing at build tools that tried to use this successfully.

**Comment by tgross35** [comment]
I only know of ninja because it was mentioned here https://github.com/rust-lang/rust/issues/80792 @dcbaker do you know any more about this?

**Comment by dcbaker** [comment]
for us —set-env alone is a massive improvement, and the main thing we need. We can use the others, but we really need the set command

**Comment by Nilstrieb** [comment]
@dcbaker have you been able to integrate it and ensure that it solves all your use cases correctly?

**Comment by sdroege** [comment]
This would be at least useful for [meson](https://mesonbuild.com/), which needs something like this because of ninja's lack of support for environment variables. Other ninja-based build systems should benefit from it in the same way.

**Comment by dcbaker** [comment]
I had planned on having a working prototype this weekend, but instead had no power send internet. Hopefully in the next day or so

**Comment by dcbaker** [comment]
Meson with this PR: https://github.com/mesonbuild/meson/pull/10030 demonstrates that `--set-env` does what we want. You can test this by running `./run_single_test.py "test cases/rust/23 env"` with a nightly (at least 1.76) and non-nightly compiler and see that it works with the nightly and not otherwise. Eventually I need to implement a fallback path that uses a wrapper script to support versions of rust without the `--set-env` option, and I need to wire this into our infrastructure to build crates as native Meson projects, but this was a straightforward usecase we can start using as soon as it's available.

File: src/doc/rustc/src/command-line-arguments.md [hunk 1]
```diff
@@ -504,3 +504,47 @@ an empty option. The file can use Unix or Windows style line endings, and must b
 encoded as UTF-8.
 
 [the JSON chapter]: json.md
+
+<a id="option-env-set"></a>
+## `env-set`: inject an environment variable
+
+This option flag allows to specify environment variables value at compile time to be
```
Comment (davidtwco) [inline src/doc/rustc/src/command-line-arguments.md:511]:
```suggestion
This option flag allows specification of environment variables at compile time for the
```

File: src/doc/rustc/src/command-line-arguments.md [hunk 2]
```diff
@@ -504,3 +504,47 @@ an empty option. The file can use Unix or Windows style line endings, and must b
...
+used by `env!` and `option_env!` macros. It also impacts `tracked_env::var` function
```
Comment (davidtwco) [inline src/doc/rustc/src/command-line-arguments.md:512]:
```suggestion
`env!` and `option_env!` macros and `tracked_env::var` function
```

File: src/doc/rustc/src/command-line-arguments.md [hunk 3]
```diff
@@ -504,3 +504,47 @@ an empty option. The file can use Unix or Windows style line endings, and must b
...
+from the `proc_macro` crate.
+
+This information will be stored in the dep-info files. For more information about
+dep-info files, take a look [here](https://doc.rust-lang.org/cargo/guide/build-cache.html#dep-info-file\
+s).
+
+When retrieving an environment variable value, the one specified by `--env-set` will take
```
Comment (davidtwco) [inline src/doc/rustc/src/command-line-arguments.md:519]:
```suggestion
When retrieving an environment variable, the value specified by `--env-set` will take
```

File: src/doc/rustc/src/command-line-arguments.md [hunk 4]
```diff
@@ -504,3 +504,47 @@ an empty option. The file can use Unix or Windows style line endings, and must b
...
+precedence. For example, if you want have `PATH=a` in your environment and pass:
```
Comment (davidtwco) [inline src/doc/rustc/src/command-line-arguments.md:520]:
```suggestion
precedence. For example, if you want have `PATH=env` in your environment and pass:
```

File: src/doc/rustc/src/command-line-arguments.md [hunk 5]
```diff
@@ -504,3 +504,47 @@ an empty option. The file can use Unix or Windows style line endings, and must b
...
+--env-set A=B --env X=12
+```
+
+and then on next compilation:
+
+```bash
+--env-set A=B
+```
+
+`X` value is different (not set) so the code will be re-compiled.
```
Comment (davidtwco) [inline src/doc/rustc/src/command-line-arguments.md:545]:
```suggestion
Crates will be fully re-compiled if `--env-set` arguments are changed.
```

**Review by davidtwco — APPROVED** [review]
Implementation LGTM, some working changes in the documentation.

**Comment by davidtwco** [comment]
Could you write a [brief stabilisation report](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#write-a-stabilization-report) describing what is being stabilised and then I'll start an MCP?

**Comment by GuillaumeGomez** [comment]
Thanks for the review @davidtwco! I added the stabilization report in the first comment of this PR.

**Comment by davidtwco** [comment]
@rfcbot fcp merge

See [stabilization report](https://github.com/rust-lang/rust/pull/119926#issue-2080303706).

**Comment by petrochenkov** [comment]
Besides `(option_)env!` and `tracked_env::var` rustc also accesses several dozens of environment variables in different parts of the compiler (starting from `PATH` and `RUSTC_BOOTSTRAP`).

None of those are tracked in depinfo.
At least some of those should supposedly be tracked (["Rebuild if deployment target changes"](https://github.com/rust-lang/rust/issues/118204)).
I suspect that some of them should also *not* be tracked to avoid unnecessary rebuilds.

I wonder how should all of them interact with `--set-env`.
Never interact? Always interact? Interact on a case-by-case basis?

@rfcbot concern other-rustc-vars

**Comment by nagisa** [comment]
I have a hard time understanding how this functionality can be useful in the form that only affects the results of the `env!`, `option_env!` and perhaps a couple other APIs but not e.g. what is inherited by the spawned child executables or returned by the usual environment variable handling APIs. In particular motivation section above also claims that some tools need an ability to override environment proper, so it would seem to me that the feature as it is proposed for stabilisation does not actually satisfy the needs of those tools?

I also have a hard time wrapping my head around why the behaviour of falling back to regular environment variables would be desirable. It seems to me that once you care enough to specify `--env-set`, you probably care about not leaking in random junk from the environment…

**Comment by xclaesse** [comment]
If I want to set an environment variable to an empty value, should we allow `rustc -Zunstable-options --env-set FOO`? That does not seems to work currently, but `rustc -Zunstable-options --env-set FOO=` works.

**Comment by Dylan-DPC** [comment]
After a discussion with the author, they have preferred this to be closed as they don't have a usecase or motivation for completing it.

**Comment by xclaesse** [comment]
@Dylan-DPC @GuillaumeGomez why is this closed? We are still waiting for this feature for Meson's rustc support.

**Comment by GuillaumeGomez** [comment]
Because it's blocked and apparently not moving any time soon.

**Comment by xclaesse** [comment]
> Besides `(option_)env!` and `tracked_env::var` rustc also accesses several dozens of environment variables in different parts of the compiler (starting from `PATH` and `RUSTC_BOOTSTRAP`).
> I wonder how should all of them interact with `--set-env`. Never interact? Always interact? Interact on a case-by-case basis?

I'm not a Rust dev, but has a Meson dev who really needs this feature, I think those other variables are unrelated.

The use-case here is really that many Rust project uses `env!` like a C program uses `#ifdef`. You typically would pass those macros as `-DFOO=BAR` but with Rust you typically do the same thing via the environment. Meson, and other build systems based on Ninja, cannot change environment while invoking rustc command. It is expected that those kind of options are passed through the CLI.

As a concrete example, serde does [`include!(concat!(env!("OUT_DIR"), "/private.rs"));`](https://github.com/serde-rs/serde/blob/b517ec9fd78446ab6282adc405bc1bf37246a4e2/serde/src/lib.rs#L260). To build serde using Meson instead of Cargo, we want to be able to run the command line `rustc --env-set OUT_DIR=...`.

Meson would basically define all those variables using CLI: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates

**Comment by GuillaumeGomez** [comment]
I don't plan to unless there is a clear intent to merge it, otherwise it'd likely just end up like this PR.

**Pull request closed without merging**